### PR TITLE
eval: Log in dev room and broadcast 

### PR DESCRIPTION
**Pull request by mia-pi-git**
Similar to /evalbattle, when ``>>`` is used, it now broadcasts into the room. /eval still does not. 
They both log into the development roomlogs. Also puts eval output that's larger than 200 characters (tentative) into a readmore details tag.

### CODE REVIEW

**Comment by scheibo** [comment]
There are absolutely cases where you want to /eval without logging to dev (for example, related to private rooms).

**Comment by mia-pi-git** [comment]
> There are absolutely cases where you want to /eval without logging to dev (for example, related to private rooms).

Perhaps it would be better to log to upperstaff, then?

**Comment by scheibo** [comment]
> Perhaps it would be better to log to upperstaff, then?

Certainly better than dev. Theres nothing in upperstaff anyway. Its still an easily bypassable hurdle though, anyone with eval access has console access and can just eval in code in a completely inauditable way anyway, so I'm not sure what the motivation here really is?

**Comment by mia-pi-git** [comment]
> > Perhaps it would be better to log to upperstaff, then?
> 
> Certainly better than dev. Theres nothing in upperstaff anyway. Its still an easily bypassable hurdle though, anyone with eval access has console access and can just eval in code in a completely inauditable way anyway, so I'm not sure what the motivation here really is?

Zarel said he'd like it. this way, and imo sometimes it's helpful to be able to retrieve eval outputs after refreshing.

**Comment by scheibo** [comment]
You can already use !eval to show what youre evaling to whatever room you choose. And >> now defaults to showing as well.

**Comment by mia-pi-git** [comment]
I don't have much more personally to say. 
@Zarel, do you have anything you want added?

**Comment by sparkychildcharlie** [comment]
hate to say it, but shouldn't this assume the logroom doesn't exist?  Didn't test it but on a blank copy of PS, wouldn't that cause a crash?  Unlike staff room, development isnt created automatically..

**Comment by Zarel** [comment]
> Its still an easily bypassable hurdle though, anyone with eval access has console access and can just eval in code in a completely inauditable way anyway, so I'm not sure what the motivation here really is?

I think you have the wrong threat model. The idea isn't to, like, protect against malicious actors who have console access. It's so if you forgot what commands you used in the past, you can look them up.

File: server/chat-commands/admin.ts [hunk 1]
```diff
@@ -929,8 +929,15 @@ export const commands: ChatCommands = {
 	async eval(target, room, user, connection) {
 		if (!this.canUseConsole()) return false;
 		if (!this.runBroadcast(true)) return;
+		const logRoom = Rooms.get('development');
```
Comment (Zarel) [inline server/chat-commands/admin.ts:932]:
```suggestion
		const logRoom = Rooms.get('upperstaff') || Rooms.get('staff');
```

File: server/chat-commands/admin.ts [hunk 2]
```diff
@@ -929,8 +929,15 @@ export const commands: ChatCommands = {
...
 
-		if (!this.broadcasting) this.sendReply(`||>> ${target}`);
+		const show = this.message.startsWith('>>') || this.message.startsWith('!eval');
+		if (show) {
+			room.add(`||>> ${target}`).update();
+			logRoom?.roomlog(`>> ${target}`);
+		} else {
+			this.sendReply(`||>> ${target}`);
+		}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:940]:
```suggestion
		} else {
			this.sendReply(`||>> ${target}`);
		}
		logRoom?.roomlog(`>> ${target}`);
```

File: server/chat-commands/admin.ts [hunk 3]
```diff
@@ -943,11 +950,29 @@ export const commands: ChatCommands = {
...
+			result = show && result.length > 200 ? (`<details class="readmore code" style="` +
+				Utils.html`display: table; tab-size: 3"><summary>${result.slice(0, 200)}</summary>` +
+				Utils.html`${result.slice(200)}</details>`
+			).replace(/\n/g, '<br>') : result.replace(/\n/g, '\n||');
+			if (show) {
+				room.add(`|html|<< ${result}`).update();
+				logRoom?.roomlog(`<< ${result}`);
+			} else {
+				this.sendReply(`|| << ${result}`);
+			}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:962]:
```suggestion
			if (show) {
				room.add(`|html|<< ${result}`).update();
			} else {
				this.sendReply(`|| << ${result}`);
			}
			logRoom?.roomlog(`<< ${result}`);
```

File: server/chat-commands/admin.ts [hunk 4]
```diff
@@ -929,8 +929,15 @@ export const commands: ChatCommands = {
 	async eval(target, room, user, connection) {
 		if (!this.canUseConsole()) return false;
 		if (!this.runBroadcast(true)) return;
+		const logRoom = Rooms.get('development');
 
-		if (!this.broadcasting) this.sendReply(`||>> ${target}`);
+		const show = this.message.startsWith('>>') || this.message.startsWith('!eval');
```
Comment (Zarel) [inline server/chat-commands/admin.ts:934]:
```suggestion
		const broadcasting = this.broadcasting || this.message.startsWith('>>');
```

(rename other uses of `show` to `broadcasting`.)

File: server/chat-commands/admin.ts [hunk 5]
```diff
@@ -943,11 +950,29 @@ export const commands: ChatCommands = {
...
+				`<details class="readmore code" style="display: table; tab-size: 3">` +
+				Utils.html`<summary>${msg.slice(0, 200)}</summary>` +
+				Utils.html`${msg.slice(200)}</details>`
+			).replace(/\n/g, '<br>') : msg.replace(/\n/g, '\n||');
+			if (show) {
+				logRoom?.roomlog(`<< ${target}`);
+				room.add(`|html|<< ${msg}`).update();
+			} else {
+				this.sendReply(`|| << ${msg}`);
+			}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:975]:
```suggestion
				room.add(`|html|<< ${msg}`).update();
			} else {
				this.sendReply(`|| << ${msg}`);
			}
			logRoom?.roomlog(`<< ${msg}`);
```

Comment (Zarel) [reply server/chat-commands/admin.ts:975]: (Notice how you used `target` instead of `msg` here.)

File: server/chat-commands/admin.ts [hunk 6]
```diff
@@ -943,11 +950,29 @@ export const commands: ChatCommands = {
 			} else {
 				result = Utils.visualize(result);
 			}
-			result = result.replace(/\n/g, '\n||');
-			this.sendReply('||<< ' + result);
+			result = show && result.length > 200 ? (`<details class="readmore code" style="` +
+				Utils.html`display: table; tab-size: 3"><summary>${result.slice(0, 200)}</summary>` +
+				Utils.html`${result.slice(200)}</details>`
+			).replace(/\n/g, '<br>') : result.replace(/\n/g, '\n||');
```
Comment (Zarel) [inline server/chat-commands/admin.ts:956]: I suspect we should probably have a `Chat` function for this (I also want to do this for inputs as well as outputs).

File: server/rooms.ts [hunk 7]
```diff
@@ -1086,6 +1086,7 @@ export class BasicChatRoom extends BasicRoom {
 	readonly creationTime: number | null;
 	readonly type: 'chat' | 'battle';
 	minorActivity: Poll | Announcement | null;
+	queuedActivity: Poll[] | null;
```
Comment (Zarel) [inline server/rooms.ts:1089]:
```suggestion
```

File: server/rooms.ts [hunk 8]
```diff
@@ -1116,6 +1117,7 @@ export class BasicChatRoom extends BasicRoom {
 		if (!options.isPersonal) this.persist = true;
 
 		this.minorActivity = null;
+		this.queuedActivity = [];
```
Comment (Zarel) [inline server/rooms.ts:1120]:
```suggestion
```

File: server/chat-commands/admin.ts (hunk 6 above)
Comment (Zarel) [reply server/chat-commands/admin.ts:956]: Also it shouldn't be at 200 characters, it should be at the first line break.

(I'll make a separate commit to cut make `Utils.visualize` puts line breaks in the right places.)

**Comment by Zarel** [comment]
> hate to say it, but shouldn't this assume the logroom doesn't exist? Didn't test it but on a blank copy of PS, wouldn't that cause a crash? Unlike staff room, development isnt created automatically..

@sparkychildcharlie Notice that the code uses `logRoom?.roomlog(` rather than `logRoom.roomlog(`. If the room doesn't exist, it just won't log. It's a pretty useful setup.

**Comment by Zarel** [comment]
@mia-pi-git Nice! This is overall pretty close to what I wanted!

File: server/chat.ts [hunk 9]
```diff
@@ -1701,6 +1701,27 @@ export const Chat = new class {
 		htmlContent = htmlContent.replace(/\n/g, '&#10;');
 		return htmlContent;
 	}
+	getReadmoreCodeBlock(str: string) {
+		const params = str.substr(+str.startsWith('\n')).split('\n');
+		if (params.length === 1 && params[0].length < 80 && !params[0].includes('```')) {
+			return `<code>${params[0]}</code>`;
+		}
```
Comment (Zarel) [inline server/chat.ts:1708]: This should be moved back out; it's a special case of `!code` that's not really relevant to other use-cases.

```suggestion
```

File: server/chat-commands/info.ts [hunk 10]
```diff
@@ -2564,34 +2564,14 @@ export const commands: ChatCommands = {
 
 	'!code': true,
 	code(target, room, user) {
-		// XXX: target is trimmed by Chat#splitMessage. Let's not add another
-		// awful hack like ! or help command keys for whether or not the target
-		// is raw for now.
-		target = this.message.substr(this.cmdToken.length + this.cmd.length + +this.message.includes(' ')).trimRight();
```
Comment (Zarel) [inline server/chat-commands/info.ts:2570]: Why did you remove this?

Comment (Zarel) [reply server/chat-commands/info.ts:2570]: This is relevant for leading whitespace (lines starting with the empty space `" "`)

Comment (mia-pi-git) [reply server/chat-commands/info.ts:2570]: i just moved it downwards a bit

File: server/chat-commands/info.ts [hunk 11]
```diff
@@ -2564,34 +2564,14 @@ export const commands: ChatCommands = {
...
-
-		const output = [];
-		let cutoff = 3;
-		for (const param of params) {
-			if (output.length < 3 && param.length > 80) cutoff = 2;
-			output.push(Utils.escapeHTML(param));
+		const code = Chat.getReadmoreCodeBlock(target);
+		if (/<code>(.+)<\/code>/.test(code)) {
+			return `/html ${code}`;
 		}
```
Comment (Zarel) [inline server/chat-commands/info.ts:2574]:
```suggestion
		const code = Chat.getReadmoreCodeBlock(target);
```

File: server/chat-commands/info.ts (hunk 10 above)
Comment (Zarel) [reply server/chat-commands/info.ts:2570]: Oh, you moved it down and removed the comment. Why?

File: server/chat-commands/info.ts (hunk 11 above)
Comment (mia-pi-git) [reply server/chat-commands/info.ts:2574]: Doing.

File: server/chat-commands/info.ts [hunk 12]
```diff
@@ -2564,34 +2564,14 @@ export const commands: ChatCommands = {
...
-		target = this.message.substr(this.cmdToken.length + this.cmd.length + +this.message.includes(' ')).trimRight();
 		if (!target) return this.parse('/help code');
 		if (target.length >= 8192) return this.errorReply("Your code must be under 8192 characters long!");
-
-		const params = target.substr(+target.startsWith('\n')).split('\n');
-		if (params.length === 1 && params[0].length < 80 && !params[0].includes('```') && this.shouldBroadcast()) {
-			return this.canTalk(`\`\`\`${params[0]}\`\`\``);
-		}
-
+		target = this.message.substr(this.cmdToken.length + this.cmd.length + +this.message.includes(' ')).trimRight();
```
Comment (Zarel) [inline server/chat-commands/info.ts:2569]:
```suggestion
		if (target.length < 80 && !target.includes('\n') && !target.includes('```') && this.shouldBroadcast()) {
			return this.canTalk(`\`\`\`${params[0]}\`\`\``);
		}

```

File: server/chat-commands/info.ts [hunk 13]
```diff
@@ -2564,34 +2564,14 @@ export const commands: ChatCommands = {
 
 	'!code': true,
 	code(target, room, user) {
-		// XXX: target is trimmed by Chat#splitMessage. Let's not add another
-		// awful hack like ! or help command keys for whether or not the target
-		// is raw for now.
-		target = this.message.substr(this.cmdToken.length + this.cmd.length + +this.message.includes(' ')).trimRight();
 		if (!target) return this.parse('/help code');
```
Comment (Zarel) [inline server/chat-commands/info.ts:2567]:
```suggestion
		// target is trimmed by Chat#splitMessage, but leading spaces can be
		// important to code block indentation.
		target = this.message.substr(this.cmdToken.length + this.cmd.length + +this.message.includes(' ')).trimRight();
		if (!target) return this.parse('/help code');
```

File: server/chat-commands/info.ts (hunk 10 above)
Comment (mia-pi-git) [reply server/chat-commands/info.ts]: Comment didn't seem necessary - as for the trimming, i originally removed it, then realized it was necessary and put it back in slightly lower.

File: server/chat-commands/info.ts [hunk 14]
```diff
@@ -2564,34 +2564,16 @@ export const commands: ChatCommands = {
...
 		target = this.message.substr(this.cmdToken.length + this.cmd.length + +this.message.includes(' ')).trimRight();
 		if (!target) return this.parse('/help code');
 		if (target.length >= 8192) return this.errorReply("Your code must be under 8192 characters long!");
-
-		const params = target.substr(+target.startsWith('\n')).split('\n');
-		if (params.length === 1 && params[0].length < 80 && !params[0].includes('```') && this.shouldBroadcast()) {
-			return this.canTalk(`\`\`\`${params[0]}\`\`\``);
+		if (target.length < 80 && !target.includes('\n') && !target.includes('```') && this.shouldBroadcast()) {
+			return this.canTalk(`\`\`\`${target}\`\`\``);
 		}
```
Comment (Zarel) [inline server/chat-commands/info.ts:2574]:
```suggestion
		}

```

File: server/chat-commands/info.ts [hunk 15]
```diff
@@ -2564,34 +2564,16 @@ export const commands: ChatCommands = {
...
-
 		if (!this.canBroadcast(true, '!code')) return;
```
Comment (Zarel) [inline server/chat-commands/info.ts:2575]:
```suggestion
		if (!this.canBroadcast(true, '!code')) return;

```

File: server/chat.ts [hunk 16]
```diff
@@ -1701,6 +1701,29 @@ export const Chat = new class {
...
+		let cutoff = 3;
+		for (const param of params) {
+			if (output.length < 3 && param.length > 80) cutoff = 2;
+			output.push(Utils.escapeHTML(param));
+		}
+
+		let code;
+		if (output.length > cutoff) {
+			code = `<div class="chat"><details class="readmore code" style="white-space: pre-wrap; display: table; tab-size: 3"><summary>` +
+				`${output.slice(0, cutoff).join('<br />')}</summary>${output.slice(cutoff).join('<br />')}</details></div>`;
```
Comment (Zarel) [inline server/chat.ts:1721]:
```suggestion
			code = `<div class="chat"><details class="readmore code" style="white-space: pre-wrap; display: table; tab-size: 3"><summary>${
				output.slice(0, cutoff).join('<br />')
			}</summary>${output.slice(cutoff).join('<br />')}</details></div>`;
```

File: server/chat.ts [hunk 17]
```diff
@@ -1701,6 +1701,29 @@ export const Chat = new class {
...
+		} else {
+			code = `<div class="chat"><code style="white-space: pre-wrap; display: table; tab-size: 3">${output.join('<br />')}</code></div>`;
```
Comment (Zarel) [inline server/chat.ts:1724]:
```suggestion
			code = `<div class="chat"><code style="white-space: pre-wrap; display: table; tab-size: 3">${
				output.join('<br />')
			}</code></div>`;
```

File: server/chat.ts [hunk 18]
```diff
@@ -1701,6 +1701,32 @@ export const Chat = new class {
...
+		for (const param of params) {
+			if (output.length < 3 && param.length > 80) cutoff = 2;
+			output.push(Utils.escapeHTML(param));
+		}
+
+		let code;
+		if (output.length > cutoff) {
+			code = `<div class="chat"><details class="readmore code" style="white-space: pre-wrap; display: table; tab-size: 3"><summary>${
+				output.slice(0, cutoff).join('<br />')
+			}</summary>${output.slice(cutoff).join('<br />')}</details></div>`;
```
Comment (Zarel) [inline server/chat.ts:1722]:
```suggestion
			}</summary>${
				output.slice(cutoff).join('<br />')
			}</details></div>`;
```

File: server/chat.ts [hunk 19]
```diff
@@ -1701,6 +1701,34 @@ export const Chat = new class {
...
 	}
+	/**
+	 * Takes a string of code and transforms it into a block of html using the details tag.
+	 * If it has a newline, will make the 3 lines the preview, and fill the rest in.
+	 * @param str string to block
+	 */
+	getReadmoreCodeBlock(str: string) {
+		const params = str.substr(+str.startsWith('\n')).split('\n');
+		const output = [];
+		let cutoff = 3;
```
Comment (Zarel) [inline server/chat.ts:1712]:
```suggestion
	getReadmoreCodeBlock(str: string, cutoff = 3) {
		const params = str.substr(+str.startsWith('\n')).split('\n');
		const output = [];
```

File: server/chat.ts [hunk 20]
```diff
@@ -1701,6 +1701,34 @@ export const Chat = new class {
...
+		for (const param of params) {
+			if (output.length < 3 && param.length > 80) cutoff = 2;
```
Comment (Zarel) [inline server/chat.ts:1714]:
```suggestion
			if (output.length < 3 && param.length > 80) cutoff = 2;
```
```suggestion
			if (output.length < cutoff && param.length > 80 && cutoff > 2) cutoff--;
```

File: server/chat-commands/admin.ts [hunk 21]
```diff
@@ -929,8 +929,15 @@ export const commands: ChatCommands = {
...
 		if (!this.runBroadcast(true)) return;
+		const logRoom = Rooms.get('upperstaff') || Rooms.get('staff');
 
-		if (!this.broadcasting) this.sendReply(`||>> ${target}`);
+		const broadcasting = this.broadcasting || this.message.startsWith('>>');
+		if (broadcasting) {
+			room.add(`||>> ${target}`).update();
+		} else {
+			this.sendReply(`||>> ${target}`);
+		}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:939]:
```suggestion
		if (this.message.startsWith('>>') && room) {
			this.broadcasting = true;
			this.broadcastToRoom = true;
		}
		this.sendReply(`|html|<table><tr><td>&raquo;</td><td>${Chat.getReadmoreCodeBlock(target)}</td></tr><table>`);
```

File: server/chat-commands/admin.ts [hunk 22]
```diff
@@ -943,11 +950,22 @@ export const commands: ChatCommands = {
...
 			}
-			result = result.replace(/\n/g, '\n||');
-			this.sendReply('||<< ' + result);
+			logRoom?.roomlog(`<< ${result}`);
+			result = broadcasting ? Chat.getReadmoreCodeBlock(result) : result.replace(/\n/g, '\n||');
+			if (broadcasting) {
+				room.add(`|html|<< ${result}`).update();
+			} else {
+				this.sendReply(`|| << ${result}`);
+			}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:958]:
```suggestion
			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(result)}</td></tr><table>`);
```

File: server/chat-commands/admin.ts [hunk 23]
```diff
@@ -943,11 +950,22 @@ export const commands: ChatCommands = {
...
-			const message = ('' + e.stack).replace(/\n *at CommandContext\.eval [\s\S]*/m, '').replace(/\n/g, '\n||');
-			this.sendReply(`|| << ${message}`);
+			let msg = ('' + e.stack).replace(/\n *at CommandContext\.eval [\s\S]*/m, '');
+			logRoom?.roomlog(`<< ${msg}`);
+			msg = broadcasting ? Chat.getReadmoreCodeBlock(msg) : msg.replace(/\n/g, '\n||');
+			if (broadcasting) {
+				room.add(`|html|<< ${msg}`).update();
+			} else {
+				this.sendReply(`|| << ${msg}`);
+			}
```
Comment (Zarel) [inline server/chat-commands/admin.ts:962]:
```suggestion
			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(msg)}</td></tr><table>`);
			logRoom?.roomlog(`<< ${msg}`);
```

File: server/chat.ts [hunk 24]
```diff
@@ -1701,6 +1701,33 @@ export const Chat = new class {
...
+				output.slice(0, cutoff).join('<br />')
+			}</summary>${
+				output.slice(cutoff).join('<br />')
+			}</details></div>`;
+		} else {
+			code = `<div class="chat"><code style="white-space: pre-wrap; display: table; tab-size: 3">${
+				output.join('<br />')
+			}</code></div>`;
+		}
+		return code;
```
Comment (Zarel) [inline server/chat.ts:1729]:
```suggestion
		if (output.length > cutoff) {
			return `<div class="chat"><details class="readmore code" style="white-space: pre-wrap; display: table; tab-size: 3"><summary>${
				output.slice(0, cutoff).join('<br />')
			}</summary>${
				output.slice(cutoff).join('<br />')
			}</details></div>`;
		} else {
			return `<div class="chat"><code style="white-space: pre-wrap; display: table; tab-size: 3">${
				output.join('<br />')
			}</code></div>`;
		}
```

File: server/chat-commands/admin.ts [hunk 25]
```diff
@@ -943,11 +949,12 @@ export const commands: ChatCommands = {
...
-			result = result.replace(/\n/g, '\n||');
-			this.sendReply('||<< ' + result);
+			logRoom?.roomlog(`<< ${result}`);
+			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(result)}</td></tr><table>`);
 		} catch (e) {
-			const message = ('' + e.stack).replace(/\n *at CommandContext\.eval [\s\S]*/m, '').replace(/\n/g, '\n||');
-			this.sendReply(`|| << ${message}`);
+			let msg = ('' + e.stack).replace(/\n *at CommandContext\.eval [\s\S]*/m, '');
+			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(msg)}</td></tr><table>`);
+			logRoom?.roomlog(`<< ${msg}`);
```
Comment (Zarel) [inline server/chat-commands/admin.ts:957]:
```suggestion
			const message = ('' + e.stack).replace(/\n *at CommandContext\.eval [\s\S]*/m, '');
			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(message)}</td></tr><table>`);
			logRoom?.roomlog(`<< ${message}`);
```

File: server/chat-commands/admin.ts [hunk 26]
```diff
@@ -943,11 +949,12 @@ export const commands: ChatCommands = {
 			} else {
 				result = Utils.visualize(result);
 			}
-			result = result.replace(/\n/g, '\n||');
-			this.sendReply('||<< ' + result);
+			logRoom?.roomlog(`<< ${result}`);
+			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(result)}</td></tr><table>`);
```
Comment (Zarel) [inline server/chat-commands/admin.ts:953]:
```suggestion
			this.sendReply(`|html|<table><tr><td>&laquo;</td><td>${Chat.getReadmoreCodeBlock(result)}</td></tr><table>`);
			logRoom?.roomlog(`<< ${result}`);
```

(All the other roomlogs are now after the `sendReply`, so we might as well make this consistent.)

File: server/chat-commands/admin.ts [hunk 27]
```diff
@@ -929,8 +929,14 @@ export const commands: ChatCommands = {
...
 		if (!this.canUseConsole()) return false;
 		if (!this.runBroadcast(true)) return;
+		const logRoom = Rooms.get('upperstaff') || Rooms.get('staff');
 
-		if (!this.broadcasting) this.sendReply(`||>> ${target}`);
+		if (this.message.startsWith('>>') && room) {
+			this.broadcasting = true;
+			this.broadcastToRoom = true;
+		}
+		this.sendReply(`|html|<table border="0" cellspacing="0" cellpadding="0"><tr><td valign="top">&gt;&gt;&nbsp;</td><td>${Chat.getReadmoreCodeBlock(target)}</td></tr><table>`);
```
Comment (Zarel) [inline server/chat-commands/admin.ts:938]: Sadly, CSS doesn't really give us an easy way to do cell styling without a `<style>` tag or anything like that. This is unfortunately still the best way to do an invisible table.

**Comment by TheImmortal** [comment]
I liked the raw output, personally. This table is unnecessary...

**Comment by Zarel** [comment]
The table is mostly useful because I often monkeypatch e.g. `Rooms.ChatRoom.prototype`. The output gets pretty huge.
